memcopy.h

```diff
@@ -342,53 +291,108 @@ static inline unsigned char *byte_memset(unsigned char *out, unsigned len) {
-            return set_bytes(out, from, dist, len);
+   Aside from better memory bus utilisation, this means that short copies
+   (COPYCHUNKSIZE bytes or fewer) will fall straight through the loop without
+   iteration, which will hopefully make the branch prediction more reliable.
+ */
+static inline unsigned char *chunkcopy_relaxed(unsigned char *out, const unsigned char *from, unsigned len) {
+    --len;
+    storechunk(out, loadchunk(from));
+    out += (len % COPYCHUNKSIZE) + 1;
+    from += (len % COPYCHUNKSIZE) + 1;
+    len /= COPYCHUNKSIZE;
+#if defined(COPYCHUNK_UNROLL)
+    Assert(COPYCHUNK_UNROLL <= 8, "Unroll code needs more unrolling");
+    switch (len % COPYCHUNK_UNROLL) {
+    case 7: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 6: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 5: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 4: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 3: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 2: copychunk(&out, &from);
+        /*FALLTHRU*/
+    case 1: copychunk(&out, &from);
     }
-
-    return chunk_memcpy(out, from, len);
+    len /= COPYCHUNK_UNROLL;
+    while (len > 0) {
+        if (COPYCHUNK_UNROLL >= 1) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 2) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 3) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 4) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 5) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 6) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 7) copychunk(&out, &from);
+        if (COPYCHUNK_UNROLL >= 8) copychunk(&out, &from);
+        len--;
+    }
+#else
+    while (len-- > 0) {
+        storechunk(out, loadchunk(from));
+        out += INFFAST_CHUNKSIZE;
+        from += INFFAST_CHUNKSIZE;
+    }
+#endif
+    return out;
 }
 
-/* Byte by byte semantics: copy LEN bytes from FROM and write them to OUT. Return OUT + LEN. */
-static inline unsigned char *chunk_copy(unsigned char *out, unsigned char *from, int dist, unsigned len) {
-    if (len < sizeof(uint64_t)) {
-        if (dist > 0)
-            return set_bytes(out, from, dist, len);
+/*
+   Behave like chunkcopy_relaxed, but avoid writing beyond of legal output.
 
+   Accepts an additional pointer to the end of safe output.  A generic safe
+   copy would use (out + len), but it's normally the case that the end of the
+   output buffer is beyond the end of the current copy, and this can still be
+   exploited.
+ */
+static inline unsigned char *chunkcopy_safe(unsigned char *out, const unsigned char *from, unsigned len,
+                                           unsigned char *limit) {
+    Assert(out + len <= limit, "chunk copy exceeds safety limit");
+    if (limit - out < COPYCHUNKSIZE) {
+        /* Output buffer is too small to write a whole COPYCHUNKSIZE chunk, so
+           do things slowly and carefully. */
         return copy_bytes(out, from, len);
```

Comment:
These don't join up.  copy_bytes() offers a safe implementation for the tail case, but it only goes up to 8, and copychunk_t could be (and is on ARM) larger than that.